### ISSUE TITLE
Make Optionalish available for use by extensions.

### ISSUE THREAD

**Issue by 6bangs** [google/auto#359]
Is there a reason [Optionalish](https://github.com/google/auto/blob/1d316fa504f96a9caa5ac93a6074c759cd6140f3/value/src/main/java/com/google/auto/value/processor/Optionalish.java) shouldn't be visible outside of the com.google.auto.value.processor package? If extensions shouldn't use those classes directly, maybe Optionalish can be moved to common. It would certainly be useful to work with optionalish types from extensions without making a copy.

**Comment by eamonnmcmanus** [comment]
Is this a problem you have run into? Could you say something about your use case? We could put a reusable Optionalish into com.google.auto.common but it would take a certain amount of design, and it's hard to do that without having a clearer picture of what people might use it for.

**Comment by 6bangs** [comment]
Sure. This stems from https://github.com/gabrielittner/auto-value-cursor/issues/47

I was looking into possible solutions and ended up doing part of what `Optionalish` does before I realized it existed (namely, determining if a type is Guava's `Optional` or Java 8's `Optional`, `OptionalDouble`, etc).

To get a little more general, the use case is for extensions to be more easily able to determine:

- Whether or not a property is optional
- The type of the value wrapped by the `Optional`
- Which code to generate for empty/absent, ofNullable/fromNullable, etc

**Comment by skylve** [comment]
I agree with @6bangs, it would add a good tool for extensions to manage optional in a proper way.
Not only [gabrielittner/auto-value-cursor](https://github.com/gabrielittner/auto-value-cursor) could benefit but [rharter/auto-value-parcel](https://github.com/rharter/auto-value-parcel), [rharter/auto-value-gson](https://github.com/rharter/auto-value-parcel) and many others.

It could be nice to also add a way to customize `Optionalish` from the project that is using it. `Option` is a common concept, and most of the implementations share a common semantic :
- empty/absent => Create empty element
- of* => Create present element with or without null-safety (not sure we need to expose both)
- isPresent => check
- get => unwrap the value

`Optionalish` could be looking for annotated classes in the project implementing an `OptionalishHelper` interface with all the common property I enumerated (guava and java8 `Optional` could be expressed with this abstraction too).

Opening `Optionalish` for extensions and making it extendible at the same time would open the door to an unified `Optional` management and keep some customization possible (for now it's simply impossible to use a different `Optional` type without writing tons of boilerplate code, that apply for auto-value itself and the extensions that do/don't support `Optional`).
